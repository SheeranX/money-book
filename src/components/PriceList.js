import React from 'react';
import Ionicon from 'react-ionicons';
import PropTypes from 'prop-types';
// 因为没涉及数据的state和生命周期，所以使用的是函数式组件
const PriceList = ({items, onModifyItem, onDeleteItem}) => {
  return (
    <ul className='list-group'>
      {items.map((item,index) => {
        return (   
        <li className='list-group-item d-flex justify-content-between p-3 align-content-center' key={index}>
          <p>
            <Ionicon className='p-2 rounded' fontSize='30px' style={{background:'#007bff'}} color={'#fff'} icon={item.category.icon}/>
          </p>
          <span className=''>{item.title}</span>
          <span>{item.category.type === 'income' ? '+' : '-'}{item.price}</span>
          <span>{item.date}</span>
          <div className='d-flex'>
            <p onClick={() => {onModifyItem(item)}} className='p-1'>
              <Ionicon className='rounded-circle p-1' fontSize='30px' style={{background:'#007bff'}} color={'#fff'} icon={'ios-create-outline'}/>
            </p>
            <p onClick={() => {onDeleteItem(item)}} className='p-1'>
              <Ionicon className='rounded-circle p-1' fontSize='30px' style={{background:'#dc3545'}} color={'#fff'} icon={'ios-trash-outline'}/>
            </p>
          </div>
        </li>)
      })}
    </ul>
  );
};
// propsType用于验证props，ts interface的思想
PriceList.propTypes = {
  items:PropTypes.array.isRequired,
  onDeleteItem: PropTypes.func.isRequired,
  onModifyItem: PropTypes.func.isRequired
};
// 用于提供默认值，个人局觉得react组件的props校验没有vue简洁
PriceList.defaultProps = {
  onDeleteItem: () => {},
  onModifyItem: () => {}
};
export default PriceList;