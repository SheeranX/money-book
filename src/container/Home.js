import React from 'react';
import MonthPicker from '../components/MonthPicker';
import PriceList from '../components/PriceList';
import TabView from '../components/TabView';
import TotalPrice from '../components/TotalPrice';
import CreateBtn from '../components/CreateBtn';
import {LIST_VIEW, getCurrentDate, CHART_VIEW} from '../util';
class Home extends React.Component{
  constructor(props) {
    super(props);
    this.state = {
      date: '',
      active: LIST_VIEW,
      list: [{
        id: '1',
        title: 'react',
        price: 200,
        income: 200,
        outcome: 400,
        date: '2019-01-01',
        category: {
          id: 1,
          name: '旅行',
          type: 'outcome',
          icon: 'ios-plane'
        }
      }]
    };
    this.handleCreateItem = this.handleCreateItem.bind(this);
  }
  calTotalPrice(arr) {
    let _outcome = arr.reduce((total, item) => {
      return total + item.outcome;
    }, 0);
    let _income = arr.reduce((total, item) => {
      return total + item.income;
    }, 0);
    return {
      outcome: _outcome,
      income: _income
    }
  }
  handleCreateItem () {
    console.log('handleCreateItem');
  }
  changeTab = (e) => {
    this.setState({active: e})
  }
  handleDeleteItem = (e) => {
    alert(e.id);
  }
  handleodifyItem = (e) => {
    alert(e.id);
  }
  render() {
    let {date, active, list} = this.state;
    const _date = getCurrentDate(date);
    return (
      <React.Fragment>
        <div className='d-flex bg-dark p-4'>
          <div className='col-6 mt-4'>
            <span className='text-white'>选择日期</span>
            <MonthPicker year={
              _date.year
            } month={
              _date.month
            }></MonthPicker>
          </div>
          <div className='col-6'>
            <TotalPrice income={this.calTotalPrice(list).income} outcome={this.calTotalPrice(list).outcome}/>
          </div>
        </div>
        <div className='my-3'>
          <TabView activeClass={this.state.active} changeTab={this.changeTab}/>
        </div>
        <div className='text-center pb-2'>
          <CreateBtn createItem={this.handleCreateItem}/>
        </div>
        {
          active === LIST_VIEW && <PriceList items={this.state.list} onDeleteItem={this.handleDeleteItem} onModifyItem={this.handleodifyItem}/>
        }
        {
          active === CHART_VIEW && '这是图表模式'
        }
      </React.Fragment>
    );
  };
}

export default Home;