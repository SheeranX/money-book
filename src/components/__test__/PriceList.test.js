import React from 'react';
import {shallow} from 'enzyme';
import PriceList from '../PriceList';
import {items} from '../../container/Home';
import Ionicon from 'react-ionicons';
const props = {
  items: items,
  onDeleteItem: jest.fn(), // 模拟点击事件
  onModifyItem: jest.fn()
};

let wrapper;
describe('test priceList component', () => {
  // 可以等待异步动作完成再触发
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props}/>);
  });
  // snapshot
  it('should render the component to snapshots', () => {
    expect(wrapper).toMatchSnapshot();
  });
  // 内容条目是否正确渲染
  it('should render the right list items', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(props.items.length);
  });
  // 图标是否正确渲染
  it('should render the right ionicon', () => {
    let icon = wrapper.find('.list-group-item').first().find(Ionicon);
    expect(icon.first().props().icon).toEqual(props.items[0].category['icon']);
  });
  // 事件是否正确触发
  it('should trigger the right click event', () => {
    const firstItem = wrapper.find('.list-group-item').first();
    firstItem.find('.operate').find('p').first().simulate('click');
    expect(props.onModifyItem).toHaveBeenCalledWith(props.items[0]);
  });
});