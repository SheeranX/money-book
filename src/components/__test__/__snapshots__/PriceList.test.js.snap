// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test priceList component should render the component to snapshots 1`] = `ShallowWrapper {}`;
